docs/crates/records.md

```diff
@@ -0,0 +1,282 @@
+# Execution records
+
+## Motivation
+
+At the end of the [preflight execution](vm.md#preflight-execution), each chip needs to generate its trace. However, only some part of the trace is usually sufficient to determine the rest. Therefore, during the instruction execution stage, each chip will gather only necessary information about the trace. This way, serial execution is sped up, and in the trace generation stage all chips can restore their traces simultaneously from the stored information.
+
+## Records
+
+For storing this minimal information, each chip defines its own _Record_ type. Most of the times, one instruction results in generating one record, which corresponds to one trace row, but there is no such hard rule, and some chips use different models. In general, how to generate records from an instruction and how to interpret the record afterwards is up to the chip.
+
+Below we discuss how the details of records generation and the required properties of the records, as well as associated types.
+
+### Record arenas
+
+An entity for storing the records is called a _Record arena_. Its one function is to allocate a memory slice for a new record. During executing an instruction in preflight execution, a chip will request the memory slice to write its record to from a record arena, and then fill it. Here is an example:
+
+```rust
+fn execute(
+    &self,
+    state: VmStateMut<F, TracingMemory, RA>,
+    instruction: &Instruction<F>,
+) -> Result<(), ExecutionError> {
+    let record: &mut PhantomRecord = state.ctx.alloc(EmptyMultiRowLayout::default());
+    let pc = *state.pc;
+    record.pc = pc;
+    record.timestamp = state.memory.timestamp;
+    let [a, b, c] = [instruction.a, instruction.b, instruction.c].map(|x| x.as_canonical_u32());
+    record.operands = [a, b, c];
+    // ...
+}
+```
+
+> [!NOTE]
+> One can see from this example that a chip does not own its record arena, which is provided to it with the state instead.
+
+### Types of record arenas
+
+We have two implementations of record arenas: `MatrixRecordArena` for generating trace on CPU, and `DenseRecordArena` for GPU. The distinction is due to the fact that, for GPU tracegen, we have to first send the records to the GPU, then generate the trace on its side, and send the trace back to CPU. This motivates the arena to pack the record densely, hence the name. On the other hand, if we don't need to send the records anywhere, then they may already represent a partially filled trace -- maybe with gaps between their elements, which will be filled during tracegen.
```

Comment:
> and send the trace back to CPU

i don't think this is correct

Reply:
Ayush is right: we **never** send buffers from device back to host except for debugging.

@shuklaayush can you fix this directly?